### ISSUE TITLE
MobileNet and Depthwise Separable Convolution

### ISSUE THREAD

**Issue by changtimwu** [changtimwu/changtimwu.github.com#70]
this has been mentioned quite a lot recently.

* [mobilenet](https://arxiv.org/pdf/1704.04861.pdf) 
mobilenet does a better job on describing depthwise separable convolution.

* [xception](https://arxiv.org/pdf/1704.04861.pdf)
Please read the _Prior Works_ section throughly.  It describes how the depthwise seperable convolution been introduced.

solid analysis:
* https://arxiv.org/abs/1608.04337v2

math:
* http://www.songho.ca/dsp/convolution/convolution2d_separable.html

**Comment by changtimwu** [comment]
zhihu's discussion is valuable
https://www.zhihu.com/question/58941804

**Comment by changtimwu** [comment]
This answer describes some history of Separable Convolution(aka group convolution)
https://stackoverflow.com/a/40927483/443016

some info about its advantage.
https://stackoverflow.com/a/37092986/443016

**Comment by changtimwu** [comment]
an simplified resnet [resnext](https://arxiv.org/abs/1611.05431) also use this techniques a lot.

**Comment by changtimwu** [comment]
very good summary
https://arxiv.org/pdf/1701.04489.pdf

**Comment by changtimwu** [comment]
another  separable convolution heavy network  [quicknet](https://arxiv.org/abs/1701.02291)
the author is quite confident at its accuracy and speed.   need comparison with mobilenet. 
studied!  you can ignore quicknet.  no comparison with mobilenet.

**Comment by changtimwu** [comment]
https://keras.io/layers/convolutional/#separableconv2d
https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d
https://www.tensorflow.org/api_docs/python/tf/nn/separable_conv2d

**Comment by changtimwu** [comment]
better mobile net summary
https://github.com/tensorflow/models/blob/master/slim/nets/mobilenet_v1.md

TF-Slim's README is a good read.  
https://github.com/tensorflow/models/blob/master/slim/README.md

google's official implementation.  It's based on slim.
https://github.com/tensorflow/models/blob/master/slim/nets/mobilenet_v1.py

It's a bit different than the tensorflow APIs we used to.  MobileNet can be defined like this.  Array of layer objects!!   as simple as keras.  
```  python
_CONV_DEFS = [
    Conv(kernel=[3, 3], stride=2, depth=32),
    DepthSepConv(kernel=[3, 3], stride=1, depth=64),
    DepthSepConv(kernel=[3, 3], stride=2, depth=128),
    DepthSepConv(kernel=[3, 3], stride=1, depth=128),
    DepthSepConv(kernel=[3, 3], stride=2, depth=256),
    DepthSepConv(kernel=[3, 3], stride=1, depth=256),
    DepthSepConv(kernel=[3, 3], stride=2, depth=512),
    DepthSepConv(kernel=[3, 3], stride=1, depth=512),
    DepthSepConv(kernel=[3, 3], stride=1, depth=512),
    DepthSepConv(kernel=[3, 3], stride=1, depth=512),
    DepthSepConv(kernel=[3, 3], stride=1, depth=512),
    DepthSepConv(kernel=[3, 3], stride=1, depth=512),
    DepthSepConv(kernel=[3, 3], stride=2, depth=1024),
    DepthSepConv(kernel=[3, 3], stride=1, depth=1024)
]
```

need time to study the two apis.  what's the difference than tf.nn. series?

``` python
 net = slim.conv2d(net, depth(conv_def.depth), conv_def.kernel,
                            stride=conv_def.stride,
                            normalizer_fn=slim.batch_norm,
                            scope=end_point)
         
net = slim.separable_conv2d(net, None, conv_def.kernel,
                             depth_multiplier=1,
                             stride=layer_stride,
                             rate=layer_rate,
                             normalizer_fn=slim.batch_norm,
                             scope=end_point)
```

docs are here.   I called it `contrib.slim` to distinguish it from `model.slim`
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/slim

**Comment by changtimwu** [comment]
mobilenet on iphone http://machinethink.net/blog/googles-mobile-net-architecture-on-iphone/?utm_source=tuicool&utm_medium=referral
his work -- `forge`  is also worthy to read.

the author wrote another article about tensorflow on ios.
http://machinethink.net/blog/tensorflow-on-ios/

**Comment by changtimwu** [comment]
interesting! you can test inference speed without real training and actual data.
https://github.com/harvitronix/keras-mobilenet/blob/master/examples/speed.py

in case tedious BatchNormalization+Activation 
https://github.com/fchollet/keras/blob/master/keras/applications/inception_v3.py#L40

**Comment by changtimwu** [comment]
Why it implements its own [depthwiseconv](https://github.com/rcmalli/keras-mobilenet/blob/master/keras_mobilenet/depthwise_conv2d.py) instead of use keras's builtin [SeparableConv2D](https://keras.io/layers/convolutional/#separableconv2d)?

rcmalli said
_Separable Convolution is already implemented in both Keras and TF but, there is no BN support after Depthwise layers (Still investigating)._


we should compare  
* [tf.nn.depthwise_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d) --[rcmalli use](https://github.com/rcmalli/keras-mobilenet/blob/master/keras_mobilenet/depthwise_conv2d.py#L122)
* [tf.nn.separable_conv2d](https://www.tensorflow.org/api_docs/python/tf/nn/separable_conv2d) -- [keras use](https://github.com/fchollet/keras/blob/5a7f6b0e74673eae86c07fc642018745443bb335/keras/backend/tensorflow_backend.py#L3178)

# DeepDog
### Excellent MobileNet tweak !! 
* remove batch normalization between depthwise and pointwise 
* use Keras's builtin `separable_conv2d`
* use `ELU` instead of `RELU`
* use `CLR` optimizer and find out the best parameters
* efficient data augmentation with https://github.com/stratospark/keras-multiprocess-image-data-generator

https://medium.com/@timanglade/how-hbos-silicon-valley-built-not-hotdog-with-mobile-tensorflow-keras-react-native-ef03260747f3

**Comment by changtimwu** [comment]
computation cost 
take the first depthwise separable convolution in mobilenet as example(layer 2,3)
<img width="327" alt="mobilenet_body_architecture" src="https://user-images.githubusercontent.com/400410/27507645-aaa54ac8-5905-11e7-98f6-8e21e46478c2.png">

depthwise separable convolution =
* a)  `3x3 depthwise convolution` for filtering 
* b)  `1x1 pointwise convolution` for combining

* M: number of input layers = 32
* N: number of output layers = 64
* Dk * Dk: kernel size= 3 * 3
* Df * Dk: input or output shape = 112 * 112

computation cost of
*  3x3 depthwise convolution: _Dk * Dk * M * Df * Df_  = 3 * 3 * 112 * 112
* 1x1 pointwise convolution: _M * N * Df * Df_ = 32 * 64 * 112 * 112

## pointwise convolution
it's a 1x1  standard convolution
_Dk * Dk * M * N * Df * Df_ = _1 * 1 * M * N * Df * Df_ = 32 * 64 * 112 * 112

**Comment by changtimwu** [comment]
Dk=3, M=512, N=512, Df=14

<img width="338" alt="hyper_cost" src="https://user-images.githubusercontent.com/400410/27508017-4a0e4048-590f-11e7-8e6f-785d355651d5.png">


## multi-adds
* standard convolution = Df * Df * M * N * Dk * Dk = 14 * 14 * 512 * 512 * 3 * 3 = 462,422,016
* Depthwise Separable Convolution = (Dk * Dk * M * Df * Df) + (M * N * Df * Df) = (3 * 3 * 512 * 14 *14 ) + ( 512 * 512 * 14 * 14)  = 52,283,392
* Depthwise Separable Convolution(alpha=0.75) =  (Dk * Dk * (M * 0.75) * Df * Df) + ( (M * 0.75) * (N * 0.75) * Df * Df) = 29,578,752
* Depthwise Separable Convolution( alpha=0.75, rho=0.714)  
``` python
(alpha,rho)=(0.75,0.714)
(Dk,M,N,Df)=(3, 512*alpha, 512*alpha, 14*rho)
(Dk * Dk * M * Df * Df) + (M * N * Df * Df) 
15079129.454591995
```

## parameters
* standard convolution: 3*3*512*512 = 2359296
* Depthwise Separable Convolution:  3*3*512 + 1*512*512 =  266752
* Depthwise Separable Convolution(alpha=0.75): 
``` python
alpha=0.75
3*3*(512*alpha) + 1*(512*alpha)*(512*alpha)
150912
```
* Depthwise Separable Convolution( alpha=0.75, rho=0.714):  same of above. number of parameters has nothing to do with input/output dimension.

**Comment by changtimwu** [comment]
把 standard convolution 的計算細節搞懂, 仔細算出

* weights, number of parameter
* memory space
* multi-adds

才有能力去 evaluate or design network architecture
squeezenet 被 reject 是很正常的,  computation cost 根本沒降下來.

**Comment by changtimwu** [comment]
mobilenet's official implementation
https://github.com/tensorflow/models/blob/master/slim/nets/mobilenet_v1.md

**Comment by changtimwu** [comment]
It worthes to read mxnet's implementation
https://github.com/dmlc/mxnet/issues/5891

**Comment by changtimwu** [comment]
tiny yolo on android.  2fps on pixel.  I think it's too slow.
https://github.com/madeye/yolo-android

**Comment by marshallixp** [comment]
mark

**Comment by ysh329** [comment]
mark

**Comment by changtimwu** [comment]
[CondenseNet](https://arxiv.org/pdf/1711.09224.pdf) -- the authors claim that CondenseNet is much more efficient than mobilenet.  According to its result table, it archives equivalent performance with only half cost of mobilenet in terms of number of parameters and operaions.
 
<img width="367" alt="condensenet_result" src="https://user-images.githubusercontent.com/400410/33424266-eb361926-d580-11e7-80ae-1d07d63b90bd.png">

**Comment by roshan-gopalakrishnan** [comment]
Can we get the weights  in between the layers of depthwise convolution and pointwise convolution in a SeparableConv2D operation ? How can we get those weights in Keras?

**Comment by changtimwu** [comment]
A research team in google found a way to quantize mobilenet to 8-bit integer(1/4 parameter space) while losing only 2% accuracy(compared to FP32).  
https://arxiv.org/abs/1712.05877

**Comment by changtimwu** [comment]
## mobilenet V2
* [paper](https://arxiv.org/abs/1801.04381)
* nice discussion on zhihu   https://www.zhihu.com/question/265709710/answer/298245276
* Shicai's did a excellent job not only on V1/V2 implementation but also good survey of implementations. 
https://github.com/shicai/MobileNet-Caffe
* Dave analyzed how V2 is evolved, where each elements of V2 is proposed from.  Good writeup https://zhuanlan.zhihu.com/p/33999416 

## pretrained models
* v1: https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet_v1.md
* v2: https://github.com/tensorflow/models/tree/master/research/slim/nets/mobilenet

**Comment by edmondja** [comment]
Does anybody know when the activation function is being used in Keras's builtin separable_conv2d ? After depthwise or after pointwise convolution ? Or both maybe ? 

@changtimwu  Why not using BN between them ?

**Comment by jayshonzs** [comment]
Hi, 

remove batch normalization between depthwise and pointwise

is there any reason to do this? is it useless to put a normalization between 2 conv?

thank you
